### PR TITLE
[Refactoring] SR-5743 Try To Force Try Refactor implementation

### PR DESCRIPTION
**Pull request by Kacper20**
Opposite refactor to [this one](https://github.com/apple/swift/pull/12063).
Behavior is as follows:
- only one try in `do/catch` statement — statement is removed, code that is contained in `do` block is untouched (except of added exclamation to `try`)
- `try` expr that isn't nested in any `do/catch` — exclamation is added
- try that is among many other tries in `do/catch` statement — exclamation is added

I was thinking a lot about doing some more work in the last case. The alternative could be to take all of the nodes that are contained in `BraceStmt`, find the node that contains (or is) the selected `TryExpr` and extract it along with all the preceding nodes out of the `do/catch` statement. This is easy to implement, but I'm not sure whether it'll not be confusing to the user. What do you think?

Resolves [SR-5743](https://bugs.swift.org/browse/SR-5743).

cc @nkcsgexi @harlanhaskins

### CODE REVIEW

File: lib/IDE/Refactoring.cpp [hunk 1]
```diff
@@ -1973,6 +1973,89 @@ bool RefactoringActionConvertToDoCatch::performChange() {
...
+  if (Contexts.size() == 0) {
+    return TryExpressionConversionInfo(*TE);
+  }
+  auto StmtNode = Contexts.back();
+  DoCatchStmt *DCStmt = dyn_cast<DoCatchStmt>(StmtNode.dyn_cast<Stmt*>());
+
+  struct TryExprCounter: public SourceEntityWalker {
+    unsigned int Count = 0;
+    bool walkToExprPre(Expr *E) {
+      if (auto *FE = dyn_cast<TryExpr>(E)) {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Here we don't need `FE`; so `isa<TryExpr>(E)` is sufficient.

File: lib/IDE/Refactoring.cpp [hunk 2]
```diff
@@ -1973,6 +1973,89 @@ bool RefactoringActionConvertToDoCatch::performChange() {
...
+  if (DCStmt && ConversionInfo.NumberOfTriesInDoCatch == 1) {
+    //It's the only try in do catch block, remove the block.
+    auto *BodyStmt = DCStmt->getBody();
+    auto *BCStmt = dyn_cast<BraceStmt>(BodyStmt);
+    auto FirstNodeStartLoc = BCStmt->getElements().front().getStartLoc();
+    auto BeforeBodyRange = CharSourceRange(
+                                      SM,
+                                      DCStmt->getStartLoc(),
+                                      FirstNodeStartLoc);
+    EditConsumer.accept(SM, BeforeBodyRange, "");
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Maybe a method in `EditorConsumer` called `remove(SM, CharSourceRange)` is cleaner than this. Could you add that method?

File: lib/IDE/Refactoring.cpp [hunk 3]
```diff
@@ -1973,6 +1973,89 @@ bool RefactoringActionConvertToDoCatch::performChange() {
...
+    auto LastNodePastEndLoc = BCStmt->getElements()
+      .back().getEndLoc().getAdvancedLocOrInvalid(1);
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: We are not sure the last token is of length `1`. Using `Lexer::getLocForEndOfToken` instead.

File: lib/IDE/Refactoring.cpp [hunk 4]
```diff
@@ -1973,6 +1973,89 @@ bool RefactoringActionConvertToDoCatch::performChange() {
...
+    auto CatchRBraceLoc = DCStmt->getEndLoc()
+      .getAdvancedLocOrInvalid(getKeywordLen(tok::r_paren));
+    auto AfterBodyRange = CharSourceRange(
+                                    SM,
+                                    LastNodePastEndLoc,
+                                    CatchRBraceLoc);
+    EditConsumer.accept(SM, AfterBodyRange, "");
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Use the newly added `remove` function here too.

**Review by nkcsgexi — COMMENTED** [review]
Looks great in general! Some minor comments 😄

**Comment by nkcsgexi** [comment]
@Kacper20 thinking more about this, I'm little concerned that this refactoring is a transition from a more recommended language usage to a less recommended one. Personally, I'm not sure this is what we want to implicitly encourage. What do you think?

**Comment by Kacper20** [comment]
Yeah, I think the same. Wasn't convinced by this but wanted to make one to learn more about the refactorings. Closing this and moving JIRA ticket to Resolved, then. I will add `remove` function in another PR.

**Pull request closed without merging**